src/main/java/com/cloudant/fdblucene/FDBIndexInput.java

```diff
@@ -1,146 +1,130 @@
-
+import java.util.concurrent.atomic.AtomicLong;
+import org.apache.lucene.store.BufferedIndexInput;
 import org.apache.lucene.store.IndexInput;
-
-import com.cloudant.fdblucene.Utils;
-
 import com.apple.foundationdb.TransactionContext;
 import com.apple.foundationdb.subspace.Subspace;
 
 /**
- * A concrete implementation of {@link IndexInput} that reads {@code pages} from
- * FoundationDB.
+ * A concrete implementation of {@link IndexInput} that reads {@code pages} from FoundationDB.
  */
-public final class FDBIndexInput extends IndexInput {
-
-    private final TransactionContext txc;
-    private final Subspace subspace;
-    private final String name;
-    private final long offset;
-    private final long length;
-    private byte[] page;
-    private long pointer;
-    private final int pageSize;
-
-    private final ReadVersionCache readVersionCache;
-
-    FDBIndexInput(final String resourceDescription, final TransactionContext txc, final Subspace subspace,
-            final String name, final long offset, final long length, final int pageSize) {
-        super(resourceDescription);
-        this.txc = txc;
-        this.subspace = subspace;
-        this.name = name;
-        this.offset = offset;
-        this.length = length;
-        page = null;
-        pointer = 0L;
-        this.pageSize = pageSize;
-        this.readVersionCache = new ReadVersionCache();
-    }
-
-    @Override
-    public void close() throws IOException {
-        // empty.
-    }
-
-    @Override
-    public long getFilePointer() {
-        return pointer;
-    }
-
-    @Override
-    public long length() {
-        return length;
+public class FDBIndexInput extends BufferedIndexInput {
+
+  private final TransactionContext txc;
+  private final Subspace subspace;
+  private final String name;
+  private final long off;
+  private final long end;
+  private final int pageSize;
+  private ReadVersionCache readVersionCache;
+
+  public FDBIndexInput(final String resourceDescription, final TransactionContext txc,
+      final Subspace subspace, final String name, final long off, final long length,
+      final int pageSize) {
+    super(resourceDescription, pageSize);
+    this.txc = txc;
+    this.subspace = subspace;
+    this.name = name;
+    this.off = off;
+    this.end = off + length;
+    this.pageSize = pageSize;
+    this.readVersionCache = new ReadVersionCache();
+  }
+
+  @Override
+  protected void readInternal(final byte[] b, final int offset, final int len) throws IOException {
+    long pos = getFilePointer() + off;
+
+    if (pos + len > end) {
+      throw new EOFException("read past EOF: " + this);
     }
 
-    @Override
-    public byte readByte() throws IOException {
-        if (pointer == length) {
-            throw new EOFException("Attempt to read past end of file");
-        }
-
-        loadPageIfNull();
-        byte result = page[FDBUtil.posToOffset(offset + pointer, pageSize)];
-        seek(this.pointer + 1);
```

Comment:
so the biggest difference from switching to BufferedIndexInput is that you are no longer utilizing `seek`. Was this the incorrect logic you were referring to?

Reply:
no, I believe the incorrect logic is in the offset/length twiddling at some point (haven't identified where exactly). for page to be null after calling loadPageIfNull we'd need to have asked for a page beyond the end of the file, which obviously shouldn't happen. This new version feels safer by design.